homeassistant/components/amcrest/__init__.py

```diff
@@ -314,97 +320,75 @@ def _start_event_monitor(
     thread.start()
 
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the Amcrest IP Camera component."""
-    hass.data.setdefault(DATA_AMCREST, {DEVICES: {}, CAMERAS: []})
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up platform from a ConfigEntry."""
+
+    name = entry.data[CONF_NAME]
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    api = AmcrestChecker(
+        hass,
+        name=name,
+        host=host,
+        port=port,
+        user=username,
+        password=password,
+    )
 
-    for device in config[DOMAIN]:
-        name: str = device[CONF_NAME]
-        username: str = device[CONF_USERNAME]
-        password: str = device[CONF_PASSWORD]
+    try:
+        serial_number = await api.async_serial_number
+    except LoginError as err:
+        raise ConfigEntryAuthFailed(f"Invalid authentication for {name}") from err
+    except AmcrestError as err:
+        raise ConfigEntryNotReady(
+            f"Unable to connect to {name} at {host}:{port}"
+        ) from err
+
+    if (
+        entry.options.get(CONF_AUTHENTICATION, DEFAULT_AUTHENTICATION)
+        == HTTP_BASIC_AUTHENTICATION
+    ):
+        authentication: aiohttp.BasicAuth | None = aiohttp.BasicAuth(username, password)
+    else:
+        authentication = None
+
+    ffmpeg_arguments: str = entry.options.get(
+        CONF_FFMPEG_ARGUMENTS, DEFAULT_FFMPEG_ARGUMENTS
+    )
+    stream_source: str = entry.options.get(CONF_STREAM_SOURCE, DEFAULT_STREAM_SOURCE)
+    resolution: str = entry.options.get(CONF_RESOLUTION, DEFAULT_RESOLUTION)
+    control_light: bool = entry.options.get(CONF_CONTROL_LIGHT, DEFAULT_CONTROL_LIGHT)
+
+    hass.data.setdefault(DOMAIN, {DEVICES: {}, CAMERAS: []})
+    hass.data[DOMAIN][DEVICES][entry.entry_id] = AmcrestDevice(
```

Comment:
Usually we store data by

`hass.data[DOMAIN][entry.entry_id]`

What you have is ok but I'd change `DEVICES` to `ENTRIES` since it looks like config entry data